### ISSUE TITLE
BF16 training use FP32 gradient

### ISSUE THREAD

**Issue by wuwuwuxxx** [pytorch/pytorch#120168]
### 🚀 The feature, motivation and pitch

If it's possible to have bfloat16 weight with float32 gradient in pytorch? 
[code here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/engine.cpp#L885-L888) try to cast grad to the same dtype with input. However, I want to use float32 gradient for bfloat15 weight directly. It can save memory and improve accuracy.
